ext/ddtrace_profiling_native_extension/collectors_thread_context.c

```diff
@@ -766,7 +774,25 @@ static struct per_thread_context *get_context_for(VALUE thread, struct thread_co
 }
 
 static void initialize_context(VALUE thread, struct per_thread_context *thread_context, struct thread_context_collector_state *state) {
-  snprintf(thread_context->thread_id, THREAD_ID_LIMIT_CHARS, "%"PRIu64" (%lu)", native_thread_id_for(thread), (unsigned long) thread_id_for(thread));
+  uint64_t native_thread_id = native_thread_id_for(thread);
+
+  if (state->linux_tid_fallback != Qnil) { // Only available when needed
+    pid_t maybe_tid = linux_tid_fallback_for(state->linux_tid_fallback, pthread_id_for(thread));
+    if (maybe_tid > 0) native_thread_id = maybe_tid;
+  }
+
+  // Why does the thread id actually contain two ids?
+  //
+  // The `native_thread_id` is the OS-level thread id, and can be used to match up a thread reported by the profiler
+  // to system-level tools, such as the task manager or even the Datadog native profiler.
+  // (It's usually the same as `Thread#native_thread_id` on modern Ruby versions.)
+  //
+  // BUT! Ruby actually reuses native threads (https://ivoanjo.me/blog/2022/11/26/ruby-reuses-native-threads/) and thus
+  // the native thread id is not enough (and could cause confusion). Thus, we also include the result of
+  // `Thread#object_id` for the thread object to provide disambiguation. The `object_id` can also come in handy when
+  // matching up profiler results to the Ruby threads in our own tests, since not all Ruby versions we test on
+  // expose `Thread#native_thread_id`.
+  snprintf(thread_context->thread_id, THREAD_ID_LIMIT_CHARS, "%"PRIu64" (%lu)", native_thread_id, (unsigned long) thread_id_for(thread));
```

Comment:
Is this string format (`native_thread_id (thread.object_id)`) free-form? I ask because I assume there is thread correlation happening the Datadog UI down the pipeline.

Reply:
It is free-form. You're right: this format will make it harder to correlate ids with other tools.

The current trade-off is that we don't need to special-case handling of ids for Ruby -- e.g. we could submit two ids as two different labels, but then the backend would need special code for Ruby to distinguish/show/render them.

My thinking is: once we have a better idea of what we would want to with ids, we'd migrate Ruby over (one thing we also don't signal to the backend is when the tid is a real one vs when it's a fallback one that the backend won't be able to correlate). This would mean the feature wouldn't work/work as well for older versions but... that's already the case today, since we've been using this format for a few releases.